pkg/proxy/iptables/proxier.go

```diff
@@ -790,7 +867,18 @@ func (proxier *Proxier) syncProxyRules() {
 		isIPv6 := utilnet.IsIPv6(svcInfo.ClusterIP())
 		protocol := strings.ToLower(string(svcInfo.Protocol()))
 		svcNameString := svcInfo.serviceNameString
-		hasEndpoints := len(proxier.endpointsMap[svcName]) > 0
+
+		allEndpoints := proxier.endpointsMap[svcName]
+		filteredEndpoints := map[proxy.Endpoint]struct{}{}
+
+		hasEndpoints := len(allEndpoints) > 0
+
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.TopologyAwareServiceRouting) {
+			for _, ep := range proxy.FilterTopologyEndpoint(proxier.nodeName, proxier.nodeLabels, svcInfo.TopologyKeys(), allEndpoints) {
+				filteredEndpoints[ep] = struct{}{}
+			}
+			hasEndpoints = len(filteredEndpoints) > 0
```

Comment:
Let me take a deeper look on this.

Reply:
If traffic from external but externalTrafficPolicy is not Local, I think it also should be routed to filtered endpoints which match topology with current node. Why not ?

Reply:
Only service with `externalTrafficPolicy:Local` and traffic from external will be be routed to filtered endpoints if gateway is enabled, this is what I'm thought about at the beginning. If you agree with this, I think no more service chain should be added, the only thing we need to concern is calculation of probability which depends on Nfiltered and Nall.

Reply:
Done. PTAL @thockin

Reply:
> If traffic from external but externalTrafficPolicy is not Local, I think it also should be routed to filtered endpoints which match topology with current node

This pushes a ton of complexity back onto LB implementations.  Consider the case of zone topology without fallback to *.  If the external LB routes to a node that is in zone X and there are no endpoints in zone X, the connection will fail.

That *MAY* be what we want (i.e. "don't do that") but I don't think we've thought that through.  I see that a lot has changes since I looked last, so I will look at the PR with fresh eyes :)